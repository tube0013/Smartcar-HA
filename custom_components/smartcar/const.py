from enum import StrEnum, auto

DOMAIN = "smartcar"
DEFAULT_NAME = "Smartcar"
API_HOST = "https://api.smartcar.com"

PLATFORMS = [
    "sensor",
    "switch",
    "lock",
    "device_tracker",
    "binary_sensor",
    "number",
]

OAUTH2_AUTHORIZE = "https://connect.smartcar.com/oauth/authorize"
OAUTH2_TOKEN = "https://auth.smartcar.com/oauth/token"  # noqa: S105
SMARTCAR_MODE = "live"

CONF_APPLICATION_MANAGEMENT_TOKEN = "application_management_token"  # noqa: S105
CONF_CLOUDHOOK = "cloudhook"


class Scope(StrEnum):
    """Scope enumeration class."""

    READ_VEHICLE_INFO = auto()
    READ_VIN = auto()
    READ_BATTERY = auto()
    READ_CHARGE = auto()
    READ_ENGINE_OIL = auto()
    READ_FUEL = auto()
    READ_LOCATION = auto()
    READ_ODOMETER = auto()
    READ_SECURITY = auto()
    READ_TIRES = auto()
    CONTROL_CHARGE = auto()
    CONTROL_SECURITY = auto()


REQUIRED_SCOPES = [
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
]

CONFIGURABLE_SCOPES = [scope for scope in Scope if scope not in REQUIRED_SCOPES]

DEFAULT_SCOPES = [
    Scope.READ_BATTERY,
    Scope.READ_CHARGE,
    Scope.READ_LOCATION,
    Scope.READ_ODOMETER,
    Scope.READ_SECURITY,
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
    Scope.CONTROL_CHARGE,
]


class EntityDescriptionKey(StrEnum):
    """EntityDescriptionKey enumeration class."""

    PLUG_STATUS = auto()
    LOCATION = auto()
    DOOR_LOCK = auto()
    CHARGE_LIMIT = auto()
    CHARGING = auto()
    BATTERY_CAPACITY = auto()
    BATTERY_LEVEL = auto()
    CHARGING_STATE = auto()
    ENGINE_OIL = auto()
    FUEL = auto()
    FUEL_PERCENT = auto()
    FUEL_RANGE = auto()
    ODOMETER = auto()
    RANGE = auto()
    TIRE_PRESSURE_BACK_LEFT = auto()
    TIRE_PRESSURE_BACK_RIGHT = auto()
    TIRE_PRESSURE_FRONT_LEFT = auto()
    TIRE_PRESSURE_FRONT_RIGHT = auto()


DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS = {
    EntityDescriptionKey.BATTERY_LEVEL,
    EntityDescriptionKey.CHARGING_STATE,
    EntityDescriptionKey.CHARGING,
    EntityDescriptionKey.DOOR_LOCK,
    EntityDescriptionKey.LOCATION,
    EntityDescriptionKey.PLUG_STATUS,
    EntityDescriptionKey.RANGE,
}
