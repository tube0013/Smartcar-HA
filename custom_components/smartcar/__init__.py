import asyncio
from functools import partial
from http import HTTPStatus
import logging

from aiohttp import ClientResponseError
from homeassistant.components import cloud, webhook
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN, CONF_WEBHOOK_ID
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryError
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.config_entry_oauth2_flow import (
    OAuth2Session,
    async_get_config_entry_implementation,
)

from . import util
from .auth import AbstractAuth
from .auth_impl import AccessTokenAuthImpl, AsyncConfigEntryAuth
from .const import API_HOST, CONF_CLOUDHOOK, DOMAIN, PLATFORMS, Scope
from .coordinator import SmartcarVehicleCoordinator
from .errors import EmptyVehicleListError, InvalidAuthError, MissingVINError
from .types import SmartcarData
from .webhooks import handle_webhook, webhook_url_from_id

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Smartcar from a config entry.

    Returns:
        If the setup was successful.

    Raises:
        ConfigEntryError: For overlapping VIN in config entries.
    """
    implementation = await async_get_config_entry_implementation(hass, entry)
    websession = async_get_clientsession(hass)
    oauth_session = OAuth2Session(hass, entry, implementation)
    auth = AsyncConfigEntryAuth(websession, oauth_session, API_HOST)
    coordinators: dict[str, SmartcarVehicleCoordinator] = {}
    entry.runtime_data = SmartcarData(auth=auth, coordinators=coordinators)
    device_registry = dr.async_get(hass)
    other_vins = vehicle_vins_in_use(hass, entry)

    for vehicle_id, details in entry.data.get("vehicles", {}).items():
        vin = details["vin"]
        make = details.get("make")
        model = details.get("model")
        year = details.get("year")

        if vin in other_vins:
            msg = f"Cannot setup multiple config entries with VIN {vin}"
            raise ConfigEntryError(msg)

        # register device
        device_registry.async_get_or_create(
            config_entry_id=entry.entry_id,
            identifiers={(DOMAIN, vin)},
            manufacturer=make,
            model=f"{model} ({year})" if model and year else model,
            name=f"{make} {model}" if make and model else f"Smartcar {vin[-4:]}",
        )
        _LOGGER.info("Registered device for VIN: %s", vin)

        # create and store coordinator
        coordinator = SmartcarVehicleCoordinator(hass, auth, vehicle_id, vin, entry)
        coordinators[vin] = coordinator
        _LOGGER.debug("Coordinator created and initial data fetched for VIN: %s", vin)

    # setup platforms before doing first refresh. this gets the entity registry
    # populated with the desired entities & allows the coordinator to determine
    # what to fetch on the first refresh. (some entities, for instance, are
    # disabled by default.)
    _LOGGER.debug("Forwarding setup to platforms: %s", PLATFORMS)
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    if CONF_WEBHOOK_ID in entry.data:
        _LOGGER.info(
            "Registering webhook at url: %s",
            (await webhook_url_from_id(hass, entry.data[CONF_WEBHOOK_ID]))[0],
        )
        webhook.async_register(
            hass,
            DOMAIN,
            entry.title,
            entry.data[CONF_WEBHOOK_ID],
            partial(handle_webhook, config_entry=entry),
        )
    else:
        _LOGGER.debug("Webhooks are not enabled")

    await asyncio.gather(
        *[async_do_first_refresh(coordinator) for coordinator in coordinators.values()]
    )

    # log stored scopes once on successful setup
    _LOGGER.info(
        "Using token with scopes: %s", entry.data.get("token", {}).get("scopes")
    )

    entry.async_on_unload(entry.add_update_listener(async_update_listener))

    return True


async def async_do_first_refresh(coordinator: SmartcarVehicleCoordinator) -> None:
    await coordinator.async_config_entry_first_refresh()
    _LOGGER.debug(
        "Coordinator created and initial data fetched for VIN: %s", coordinator.vin
    )


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry.

    Returns:
        If the unload was successful.
    """
    _LOGGER.info("Unloading Smartcar entry %s", entry.entry_id)
    if CONF_WEBHOOK_ID in entry.data:
        webhook.async_unregister(hass, entry.data[CONF_WEBHOOK_ID])
    return bool(await hass.config_entries.async_unload_platforms(entry, PLATFORMS))


async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Cleanup when entry is removed."""
    if CONF_WEBHOOK_ID in entry.data and (
        cloud.async_active_subscription(hass) or entry.data.get(CONF_CLOUDHOOK, False)
    ):
        try:
            _LOGGER.debug(
                "Removing Smartcar cloudhook (%s)", entry.data[CONF_WEBHOOK_ID]
            )
            await cloud.async_delete_cloudhook(hass, entry.data[CONF_WEBHOOK_ID])
        except cloud.CloudNotAvailable:
            pass


async def async_update_listener(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> None:
    """Handle options update."""
    await hass.config_entries.async_reload(entry.entry_id)


async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
    _LOGGER.debug(
        "Migrating configuration from version %s.%s",
        config_entry.version,
        config_entry.minor_version,
    )

    # prevent rollbacks
    if config_entry.version > 2:
        return False

    if config_entry.version == 1:
        old_data = config_entry.data
        session = async_get_clientsession(hass)
        token = old_data[CONF_TOKEN]
        access_token = token[CONF_ACCESS_TOKEN]
        scopes = token["scope"].split(" ")
        auth = AccessTokenAuthImpl(session, access_token, API_HOST)

        # copy old data & remove old keys
        new_data = {**old_data}
        new_data[CONF_TOKEN] = {**old_data[CONF_TOKEN]}
        new_data[CONF_TOKEN].pop("scope", None)

        await populate_entry_data(new_data, auth, scopes)

        old_vehicle_ids = set(old_data.get("vehicles", {}).keys())
        new_vehicle_ids = set(new_data["vehicles"].keys())

        # limit the vehicles in the config entry to whatever was in the previous
        # entry even if the API is returning new items.
        if old_vehicle_ids:
            for vehicle_id in new_vehicle_ids:
                if vehicle_id not in old_vehicle_ids:
                    new_data["vehicles"].pop(vehicle_id, None)

        # ensure all previously accessible vehicles are still accessible.
        inaccessible_vehicle_ids = [
            vehicle_id
            for vehicle_id in old_vehicle_ids
            if vehicle_id not in new_vehicle_ids
        ]

        if inaccessible_vehicle_ids:
            _LOGGER.error(
                "Vehicle(s) are no longer accessible via the API: %s",
                inaccessible_vehicle_ids,
            )
            return False

        hass.config_entries.async_update_entry(
            config_entry,
            unique_id=util.unique_id_from_entry_data(new_data),
            data=new_data,
            version=2,
            minor_version=0,
        )

    _LOGGER.debug(
        "Migration to configuration version %s.%s successful",
        config_entry.version,
        config_entry.minor_version,
    )

    return True


def vehicle_vins_in_use(
    hass: HomeAssistant, config_entry: ConfigEntry = None
) -> set[str]:
    return {
        vehicle["vin"]
        for other_entry in hass.config_entries.async_entries(DOMAIN)
        for vehicle in other_entry.data.get("vehicles", {}).values()
        if not config_entry or other_entry.unique_id != config_entry.unique_id
    }


async def populate_entry_data(
    data: dict,
    auth: AbstractAuth,
    scopes: list[Scope],
) -> None:
    """Populate config entry data during initial creation or migration."""
    _inject_requested_scopes_into_entry_data(data, scopes)

    await _store_all_vehicles(data, auth)


def _inject_requested_scopes_into_entry_data(data: dict, scopes: list[Scope]) -> None:
    """Inject selected scopes into stored token data."""
    data.setdefault("token", {})["scopes"] = scopes


async def _store_all_vehicles(
    data: dict,
    auth: AbstractAuth,
) -> None:
    """Fetch and store data for all vehicles in config entry data.

    Raises:
        EmptyVehicleListError: If no vehicles are found.
        InvalidAuthError: If the request cannot be authorized.
        ClientResponseError: If there is a request error.
    """

    _LOGGER.info("Fetching Smartcar vehicle IDs...")

    data["vehicles"] = {}

    try:
        vehicle_list_resp = await auth.request(
            "get",
            "vehicles",
        )
        vehicle_list_resp.raise_for_status()
        vehicle_list_data = await vehicle_list_resp.json()
        vehicle_ids = vehicle_list_data.get("vehicles", [])
    except ClientResponseError as err:
        if err.status == HTTPStatus.UNAUTHORIZED:
            msg = f"Auth error fetching vehicle list: {err.status}"
            raise InvalidAuthError(msg) from err
        raise

    _LOGGER.info("Found %s vehicle IDs", len(vehicle_ids))

    if not vehicle_ids:
        raise EmptyVehicleListError

    await asyncio.gather(
        *[_store_vehicle_details(data, auth, vid) for vid in vehicle_ids]
    )


async def _store_vehicle_details(
    data: dict,
    auth: AbstractAuth,
    vehicle_id: str,
) -> None:
    """Fetch and store data for a single vehicle.

    Raises:
        MissingVINError: If the VIN is not available.
        InvalidAuthError: If the request cannot be authorized.
        ClientResponseError: If there is a request error.
    """

    try:
        _LOGGER.debug("Fetching VIN for vehicle ID: %s", vehicle_id)
        vin_resp = await auth.request(
            "get",
            f"vehicles/{vehicle_id}/vin",
        )
        vin_resp.raise_for_status()
        vin_data = await vin_resp.json()
        vin = vin_data.get("vin")

        if not vin:
            msg = f"No VIN for vehicle {vehicle_id}"
            raise MissingVINError(msg)

        data["vehicles"][vehicle_id] = {
            "vin": vin,
        }

        _LOGGER.debug("Fetching attributes for vehicle ID: %s", vehicle_id)
        attr_resp = await auth.request(
            "get",
            f"vehicles/{vehicle_id}",
        )
        attr_resp.raise_for_status()
        vehicle_info = await attr_resp.json()
        make = vehicle_info.get("make")
        model = vehicle_info.get("model")
        year = vehicle_info.get("year")

        data["vehicles"][vehicle_id].update(
            {
                "make": make,
                "model": model,
                "year": year,
            }
        )
    except ClientResponseError as err:
        if err.status == HTTPStatus.UNAUTHORIZED:
            msg = f"Auth error [{err.status}] during vehicle setup"
            raise InvalidAuthError(msg) from err
        raise
