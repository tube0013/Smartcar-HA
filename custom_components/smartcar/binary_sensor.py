from dataclasses import dataclass
import logging

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import EntityDescriptionKey
from .coordinator import SmartcarVehicleCoordinator
from .entity import SmartcarEntity, SmartcarEntityDescription

_LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True, kw_only=True)
class SmartcarBinarySensorDescription(
    BinarySensorEntityDescription, SmartcarEntityDescription
):
    """Class describing Smartcar binary sensor entities."""


SENSOR_TYPES: tuple[BinarySensorEntityDescription, ...] = (
    SmartcarBinarySensorDescription(
        key=EntityDescriptionKey.PLUG_STATUS,
        name="Charging Cable Plugged In",
        value_key_path="charge-ischargingcableconnected.value",
        device_class=BinarySensorDeviceClass.PLUG,
    ),
)


async def async_setup_entry(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    coordinators: dict[str, SmartcarVehicleCoordinator] = (
        entry.runtime_data.coordinators
    )
    entities = [
        SmartcarBinarySensor(coordinator, description)
        for coordinator in coordinators.values()
        for description in SENSOR_TYPES
        if coordinator.is_scope_enabled(description.key, verbose=True)
    ]
    _LOGGER.info("Adding %s Smartcar binary sensor entities", len(entities))
    async_add_entities(entities)


class SmartcarBinarySensor(SmartcarEntity[bool, bool], BinarySensorEntity):
    """Binary sensor entity for plugged in status."""

    _attr_has_entity_name = True

    @property
    def is_on(self) -> bool:
        return self._extract_value()
