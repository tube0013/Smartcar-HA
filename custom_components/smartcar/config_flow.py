from __future__ import annotations

from collections.abc import Mapping
import logging
from typing import Any, cast

from aiohttp import ClientConnectorError, ClientError
from homeassistant.components import cloud, webhook
from homeassistant.config_entries import (
    SOURCE_REAUTH,
    ConfigEntry,
    ConfigFlowResult,
    OptionsFlow,
)
from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN, CONF_WEBHOOK_ID
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.config_entry_oauth2_flow import AbstractOAuth2FlowHandler
from homeassistant.helpers.selector import (
    TextSelector,
    TextSelectorConfig,
    TextSelectorType,
)
import voluptuous as vol

from . import populate_entry_data, vehicle_vins_in_use
from .auth_impl import AccessTokenAuthImpl
from .const import (
    API_HOST,
    CONF_APPLICATION_MANAGEMENT_TOKEN,
    CONF_CLOUDHOOK,
    CONFIGURABLE_SCOPES,
    DEFAULT_NAME,
    DEFAULT_SCOPES,
    DOMAIN,
    REQUIRED_SCOPES,
    SMARTCAR_MODE,
    Scope,
)
from .errors import EmptyVehicleListError, InvalidAuthError, MissingVINError
from .util import unique_id_from_entry_data, vins_from_entry_data
from .webhooks import webhook_url_from_id

_LOGGER = logging.getLogger(__name__)

CONF_USE_WEBHOOKS = "use_webhooks"

GENERAL_CONFIGURATION_SCHEMA = {
    vol.Required(CONF_USE_WEBHOOKS, default=True): bool,
    vol.Optional(CONF_APPLICATION_MANAGEMENT_TOKEN): TextSelector(
        config=TextSelectorConfig(type=TextSelectorType.TEXT)
    ),
}
BASE_DESCRIPTION_PLACEHOLDERS = {
    "webhook_url": "webhooks-not-enabled",
    "smartcar_url": "https://dashboard.smartcar.com/configuration",
    "docs_url": "https://github.com/tube0013/Smartcar-HA/#webhooks",
}


def _validate_general_configuration_input(
    user_input: dict[str, Any],
    errors: dict[str, str],
) -> None:
    use_webhooks = user_input[CONF_USE_WEBHOOKS]
    management_token = user_input.get(CONF_APPLICATION_MANAGEMENT_TOKEN)

    if use_webhooks and not management_token:
        errors[CONF_APPLICATION_MANAGEMENT_TOKEN] = "no_management_token"

    if not use_webhooks and management_token:
        errors["base"] = "extraneous_management_token"

    if not management_token:
        user_input.pop(CONF_APPLICATION_MANAGEMENT_TOKEN, None)


def _add_dynamic_values_to_entry_data(
    data: dict[str, Any],
) -> dict[str, Any]:
    return (
        {
            **data,
            CONF_USE_WEBHOOKS: bool(data.get(CONF_APPLICATION_MANAGEMENT_TOKEN)),
        }
        if data
        else data
    )


class SmartcarOAuth2FlowHandler(AbstractOAuth2FlowHandler, domain=DOMAIN):  # type: ignore[call-arg]
    """Config flow to handle Smartcar OAuth2 authentication."""

    DOMAIN = DOMAIN
    VERSION = 2
    MINOR_VERSION = 0
    scope_data: dict[str, Any] | None = None

    @staticmethod
    @callback
    def async_get_options_flow(
        config_entry: ConfigEntry,  # noqa: ARG004
    ) -> OptionsFlow:
        """Get the options flow for this handler.

        Returns:
            The options flow.
        """
        return SmartcarOptionsFlow()

    @property
    def logger(self) -> logging.Logger:
        return _LOGGER

    @property
    def extra_authorize_data(self) -> dict[str, Any]:
        """Extra data that needs to be appended to the authorize url."""

        return {
            "mode": SMARTCAR_MODE,
            "scope": " ".join(self.requested_scopes),
        }

    def _initial_data(self) -> dict[str, Any]:
        return self._get_reauth_entry().data if self.source == SOURCE_REAUTH else {}

    @property
    def selected_scopes(self) -> list[Scope]:
        assert self.scope_data

        return sorted(
            [
                cast("Scope", scope)
                for scope, selected in self.scope_data.items()
                if selected
            ]
        )

    @property
    def requested_scopes(self) -> list[Scope]:
        return REQUIRED_SCOPES + self.selected_scopes

    async def async_step_user(
        self,
        user_input: dict[str, Any] | None = None,
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user.

        Returns:
            The config flow result.
        """
        errors: dict[str, str] = {}
        description_placeholders: dict[str, str] = {**BASE_DESCRIPTION_PLACEHOLDERS}

        if user_input is not None:
            user_input = {**user_input}
            _validate_general_configuration_input(user_input, errors)
            self.entry_data = {**user_input}
            self.entry_data.pop(CONF_USE_WEBHOOKS, None)

        if user_input is not None and not errors:
            return await self.async_step_scopes()

        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(
                vol.Schema(GENERAL_CONFIGURATION_SCHEMA),
                _add_dynamic_values_to_entry_data(self._initial_data())
                if user_input is None
                else user_input,
            ),
            errors=errors,
            last_step=False,
            description_placeholders=description_placeholders,
        )

    async def async_step_scopes(
        self,
        user_input: dict[str, Any] | None = None,
    ) -> ConfigFlowResult:
        """Handle the scopes selection step.

        Returns:
            The config flow result.
        """
        errors: dict[str, str] = {}

        if user_input is not None:
            self.scope_data = user_input

            if self.selected_scopes:
                return await super().async_step_user()
            errors["base"] = "no_scopes"

        return self.async_show_form(
            step_id="scopes",
            data_schema=self.add_suggested_values_to_schema(
                vol.Schema(
                    {
                        vol.Optional(str(scope), default=scope in DEFAULT_SCOPES): bool
                        for scope in CONFIGURABLE_SCOPES
                    }
                ),
                dict.fromkeys(
                    self._initial_data().get(CONF_TOKEN, {}).get("scopes", []), True
                )
                if user_input is None
                else user_input,
            ),
            errors=errors,
            last_step=False,
        )

    async def async_step_reauth(
        self,
        entry_data: Mapping[str, Any],  # noqa: ARG002
    ) -> ConfigFlowResult:
        """Perform reauth upon an API authentication error.

        Returns:
            The config flow result.
        """
        self.entry_data = {**self._initial_data()}
        return await self.async_step_reauth_confirm()

    async def async_step_reauth_confirm(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Dialog that informs the user that reauth is required.

        Returns:
            The config flow result.
        """
        if user_input is None:
            return self.async_show_form(step_id="reauth_confirm")
        return await self.async_step_scopes()

    async def async_oauth_create_entry(self, data: dict) -> ConfigFlowResult:
        session = async_get_clientsession(self.hass)
        token = data[CONF_TOKEN][CONF_ACCESS_TOKEN]
        auth = AccessTokenAuthImpl(session, token, API_HOST)
        data = {**self.entry_data, **data}
        data.pop(CONF_USE_WEBHOOKS, None)
        description_placeholders = {**BASE_DESCRIPTION_PLACEHOLDERS}

        try:
            await populate_entry_data(
                data,
                auth,
                self.requested_scopes,
            )
        except EmptyVehicleListError:
            _LOGGER.exception("No vehicles returned")
            return self.async_abort(reason="no_vehicles")
        except MissingVINError:
            _LOGGER.exception("Missing vehicle VIN")
            return self.async_abort(reason="unknown")
        except InvalidAuthError:
            _LOGGER.exception("Failed to authenticate")
            return self.async_abort(reason="invalid_access_token")
        except (ClientConnectorError, ClientError):
            _LOGGER.exception("Failed to fetch vehicles")
            return self.async_abort(reason="cannot_connect")

        await self.async_set_unique_id(unique_id_from_entry_data(data))

        other_vins = vehicle_vins_in_use(
            self.hass,
            self._get_reauth_entry() if self.source == SOURCE_REAUTH else None,
        )
        duplicate_vins = [
            details["vin"]
            for details in data.get("vehicles", {}).values()
            if details["vin"] in other_vins
        ]

        if duplicate_vins:
            return self.async_abort(
                reason="duplicate_vehicles",
                description_placeholders={"vins": duplicate_vins},
            )

        if self.source == SOURCE_REAUTH:
            reauth_entry = self._get_reauth_entry()

            self._abort_if_unique_id_mismatch(
                reason="wrong_vehicles",
                description_placeholders={
                    "vins": vins_from_entry_data(self._initial_data())
                },
            )

            return self.async_update_reload_and_abort(
                reauth_entry, data={**self._initial_data(), **data}
            )

        self._abort_if_unique_id_configured()

        # populate webhook details
        if CONF_APPLICATION_MANAGEMENT_TOKEN in data:
            try:
                webhook_id, webhook_url, cloudhook = await _get_webhook_details(
                    self.hass
                )
            except cloud.CloudNotConnected:
                return self.async_abort(reason="cloud_not_connected")
            data = {
                **data,
                CONF_WEBHOOK_ID: webhook_id,
                CONF_CLOUDHOOK: cloudhook,
            }
            description_placeholders = {
                **description_placeholders,
                "webhook_url": webhook_url,
            }

        return self.async_create_entry(
            title=DEFAULT_NAME,
            data=data,
            description_placeholders=description_placeholders,
        )


class SmartcarOptionsFlow(OptionsFlow):
    """Handle a option flow."""

    def _initial_data(self) -> dict[str, Any]:
        result: dict[str, Any] = self.config_entry.data
        return result

    async def async_step_init(
        self,
        user_input: dict[str, Any] | None = None,
    ) -> ConfigFlowResult:
        """Handle options flow.

        Returns:
            The config flow result.
        """
        entry_data = {**self.config_entry.data}
        errors: dict[str, str] = {}
        description_placeholders: dict[str, str] = {**BASE_DESCRIPTION_PLACEHOLDERS}

        if user_input is not None:
            user_input = {**user_input}
            _validate_general_configuration_input(user_input, errors)

        if user_input is not None and not errors:
            entry_data.pop(CONF_APPLICATION_MANAGEMENT_TOKEN, None)
            entry_data.update(user_input)
            entry_data.pop(CONF_USE_WEBHOOKS, None)

        # always try to populate webhook details since the url is used in the
        # description placeholders. (the entry_data will not be saved if there
        # were errors.)
        if entry_data.get(CONF_APPLICATION_MANAGEMENT_TOKEN):
            try:
                webhook_id, webhook_url, cloudhook = await _get_webhook_details(
                    self.hass, entry_data.get(CONF_WEBHOOK_ID)
                )
            except cloud.CloudNotConnected:
                return self.async_abort(reason="cloud_not_connected")
            entry_data = {
                **entry_data,
                CONF_WEBHOOK_ID: webhook_id,
                CONF_CLOUDHOOK: cloudhook,
            }
            description_placeholders = {
                **description_placeholders,
                "webhook_url": webhook_url,
            }
        else:
            entry_data.pop(CONF_WEBHOOK_ID, None)
            entry_data.pop(CONF_CLOUDHOOK, None)

        if user_input is not None and not errors:
            self.hass.config_entries.async_update_entry(
                self.config_entry,
                data=entry_data,
            )
            return self.async_create_entry(
                data={},
                description_placeholders=description_placeholders,
            )

        return self.async_show_form(
            step_id="init",
            data_schema=self.add_suggested_values_to_schema(
                vol.Schema(GENERAL_CONFIGURATION_SCHEMA),
                _add_dynamic_values_to_entry_data(
                    self._initial_data(),
                )
                if user_input is None
                else user_input,
            ),
            errors=errors,
            last_step=True,
            description_placeholders=description_placeholders,
        )


async def _get_webhook_details(
    hass: HomeAssistant, webhook_id: str | None = None
) -> tuple[str, str, bool]:
    if webhook_id is None:
        webhook_id = webhook.async_generate_id()
    return (webhook_id, *(await webhook_url_from_id(hass, webhook_id)))
