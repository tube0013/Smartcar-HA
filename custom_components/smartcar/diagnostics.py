"""Diagnostics support for Smartcar."""

from typing import Any, cast

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ACCESS_TOKEN,
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
    CONF_WEBHOOK_ID,
)
from homeassistant.core import HomeAssistant

from .coordinator import SmartcarVehicleCoordinator
from .webhooks import webhook_url_from_id

CONF_REFRESH_TOKEN = "refresh_token"  # noqa: S105
CONF_VIN = "vin"

TO_REDACT = {
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
    CONF_ACCESS_TOKEN,
    CONF_REFRESH_TOKEN,
    CONF_VIN,
}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinators: dict[str, SmartcarVehicleCoordinator] = (
        entry.runtime_data.coordinators
    )

    return cast(
        "dict[str, Any]",
        async_redact_data(
            {
                "entry": entry.as_dict(),
                "webhook_url": (
                    await webhook_url_from_id(hass, entry.data[CONF_WEBHOOK_ID])
                )[0]
                if CONF_WEBHOOK_ID in entry.data
                else None,
                "data": {
                    coordinator_name: coordinator.data
                    for coordinator_name, coordinator in coordinators.items()
                },
            },
            TO_REDACT,
        ),
    )
