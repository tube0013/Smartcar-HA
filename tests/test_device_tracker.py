"""Test device trackers."""

from typing import Any

from homeassistant.core import HomeAssistant, State
from homeassistant.helpers.restore_state import STORAGE_KEY as RESTORE_STATE_KEY
import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_mock_restore_state_shutdown_restart,
    mock_restore_cache_with_extra_data,
)
from syrupy.assertion import SnapshotAssertion

from custom_components.smartcar.const import (
    DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS,
    EntityDescriptionKey,
)

from . import setup_added_integration, setup_integration

RESTORE_STATE_PARAMETRIZE_ARGS = [
    (
        "entity_id",
        "stored_data",
        "device_tracker_state",
        "coordinator_data",
    ),
    [
        (
            "device_tracker.vw_id_4_location",
            {"raw_value": {"latitude": 37.4292, "longitude": 122.1381}},
            "not_home",
            {
                "location-preciselocation": {
                    "value": {"latitude": 37.4292, "longitude": 122.1381}
                }
            },
        ),
    ],
]
RESTORE_STATE_PARAMETRIZE_IDS = [
    "location",
]


@pytest.mark.parametrize(
    *RESTORE_STATE_PARAMETRIZE_ARGS,
    ids=RESTORE_STATE_PARAMETRIZE_IDS,
)
@pytest.mark.usefixtures("enable_specified_entities")
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
@pytest.mark.parametrize(
    "enabled_entities",
    [DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS | {EntityDescriptionKey.LOCATION}],
)
async def test_restore_device_tracker_save_state(
    hass: HomeAssistant,
    hass_storage: dict[str, Any],
    mock_config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
    vehicle_attributes: dict,
    entity_id: str,
    stored_data: dict,
    device_tracker_state: Any,
    coordinator_data: dict,
) -> None:
    """Test saving device tracker/coordinator state."""

    await setup_integration(hass, mock_config_entry)

    coordinator = mock_config_entry.runtime_data.coordinators[vehicle_attributes["vin"]]
    coordinator.data = coordinator_data

    await async_mock_restore_state_shutdown_restart(hass)  # trigger saving state

    stored_entity_data = [
        item["extra_data"]
        for item in hass_storage[RESTORE_STATE_KEY]["data"]
        if item["state"]["entity_id"] == entity_id
    ]

    assert stored_entity_data[0] == stored_data
    assert stored_entity_data == snapshot


@pytest.mark.parametrize(
    *RESTORE_STATE_PARAMETRIZE_ARGS,
    ids=RESTORE_STATE_PARAMETRIZE_IDS,
)
@pytest.mark.usefixtures("enable_specified_entities")
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
@pytest.mark.parametrize(
    "enabled_entities",
    [DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS | {EntityDescriptionKey.LOCATION}],
)
async def test_restore_state(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    vehicle_attributes: dict,
    entity_id: str,
    stored_data: dict,
    device_tracker_state: Any,
    coordinator_data: dict,
) -> None:
    """Test device tracker restore state."""

    mock_restore_cache_with_extra_data(
        hass,
        (
            (
                State(
                    entity_id,
                    "does-not-matter-for-this-test",
                ),
                stored_data,
            ),
        ),
    )

    mock_config_entry.add_to_hass(hass)
    hass.config_entries.async_update_entry(
        mock_config_entry,
        pref_disable_polling=True,
    )

    await setup_added_integration(hass, mock_config_entry)

    coordinator = mock_config_entry.runtime_data.coordinators[vehicle_attributes["vin"]]
    state = hass.states.get(entity_id)
    assert state
    assert state.state == device_tracker_state
    assert coordinator.data == coordinator_data
