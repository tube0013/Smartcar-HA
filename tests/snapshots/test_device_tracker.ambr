# serializer version: 1
# name: test_restore_device_tracker_save_state[enabled_entities0-vw_id_4-location]
  list([
    dict({
      'raw_value': dict({
        'latitude': 37.4292,
        'longitude': 122.1381,
      }),
    }),
  ])
# ---
